### PR TITLE
Simplify inventory checks

### LINKED ISSUE #210: multiple checks for local distro?
It looks like `NodeDistro::remote` and `NodeDistro::fetch` both check for a local distro for the requested version. This seems fishy -- is this checking for the same thing twice?

It seems like the check for a locally stored version should be abstracted out if possible. And `NodeDistro::fetch` is too long and should be broken up.

### PR DESCRIPTION
**Pull request by charlespierce**
Closes #210 

Info
-----
Our current implementation of `Inventory` has a few problems with how it collects data:
* For Node & Yarn, it looks at the Tarball cache to determine if a package is available, not the actual tool image. This means that if either the cached tarball or the image are removed, Volta winds up in an irrecoverable state.
* While the `Inventory` object is loaded lazily, once we load it, we load everything. So even if all we need to know is, "Do we have one specific version of Node available?", we have to catalogue all the Node, Yarn, and Package versions before determining the answer to the question. Most often on the hot path, we only need to know about a couple specific tools, and not the whole of the inventory.
* Because the Inventory was stateful, we needed to keep it in sync as we were doing fetches, which opened to the possibility of it getting out of sync at some point.

Changes
-----
* Removed the `Inventory` and `LazyInventory` structs, replacing them with functions that answer a specific question.
* Added `node_available`, `yarn_available`, and `package_available` functions that check for a single, specific version of Node, Yarn, or a Package, respectively.
* Added `node_versions`, `yarn_versions`, and `package_configs` functions, which return a Set of all available versions for each tool (similar to how the previous `Collection` objects were built).
* Updated the logic for Node and Yarn to look at the image directory, instead of the cached tarballs, to determine if a tool is installed.
* Renamed `fetch_internal` in all of the Tool implementations to `ensure_fetched`, which more accurately reflects what the method does.

Notes
-----
* The logic for determining the list of installed Packages was left unchanged, as it depends on the existence of `PackageConfig` files, instead of the image directories (since a package can have multiple binaries).
* Since we are no longer caching the entire inventory, there is the possibility of duplicate IO calls. However, only `volta list` is currently using the full list checks, and everywhere else is only doing single `$tool_available` calls. Currently there aren't any paths that need multiple calls, but perhaps we should try to instrument that in tests so we don't get regressions going forward.
* Also cleaned up some unneeded `.clone()` calls in tests that produced Clippy warnings.

### CODE REVIEW

**Comment by charlespierce** [comment]
Super non-scientific, anecdotal test: Running `node --version` 200 times on my Mac gives me an average runtime of `16.7 ms +- 1.3 ms`, while doing the same with this change gives me an average runtime of `14.5 ms +- 1.2 ms`, so this seems to give a significant (though small in an absolute sense) improvement to the hot path execution.

**Comment by rwjblue** [comment]
Awesome, thank you for testing!

**Comment by chriskrycho** [comment]
For a slightly more significant test—

Before:

```sh
$ hyperfine "node --version" --warmup 10
Benchmark #1: node --version
  Time (mean ± σ):      13.2 ms ±   0.7 ms    [User: 6.4 ms, System: 4.5 ms]
  Range (min … max):    11.8 ms …  16.1 ms    172 runs
```

After:

```sh
hyperfine "node --version" --warmup 100
Benchmark #1: node --version
  Time (mean ± σ):      13.6 ms ±   0.8 ms    [User: 6.4 ms, System: 4.8 ms]
  Range (min … max):    11.7 ms …  15.4 ms    162 runs
```

Doesn't seem like it makes a meaningful difference at least with that.

---

Methodology: ran it with the original, then copied the original out of its place and copied the new release build into its place.

**Comment by chriskrycho** [comment]
Follow-up: my methodology was wrong; didn't have the relevant files in place. After doing it correctly (`./dev/unix/volta-install.sh --release`):

```sh
hyperfine "node --version" --warmup 100
Benchmark #1: node --version
  Time (mean ± σ):      12.1 ms ±   0.3 ms    [User: 5.6 ms, System: 4.0 ms]
  Range (min … max):    11.6 ms …  13.4 ms    201 runs
```

Confirms @charlespierce's results almost exactly!

File: crates/volta-core/src/inventory.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    volta_home().map(|home| {
+        home.node_image_root_dir()
+            .join(version.to_string())
+            .exists()
+    })
+}
+
+/// Collects a set of all Node versions fetched on the local machine
+pub fn node_versions() -> Fallible<BTreeSet<Version>> {
+    volta_home().and_then(|home| read_versions(home.node_image_root_dir()))
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/inventory.rs:29]: That _is_ nice, but has an issue with getting the correct version of `VoltaHome` in scope for the 2nd line. It's likely reasonable to have the `volta_layout` crate re-export the current "state-of-the-art" layout at the top level, but outside the scope of this PR.

File: crates/volta-core/src/inventory.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+/// Checks if a given Yarn version image is available on the local machine
+pub fn yarn_available(version: &Version) -> Fallible<bool> {
+    volta_home().map(|home| home.yarn_image_dir(&version.to_string()).exists())
+}
+
+/// Collects a set of all Yarn versions fetched on the local machine
+pub fn yarn_versions() -> Fallible<BTreeSet<Version>> {
+    volta_home().and_then(|home| read_versions(home.yarn_image_root_dir()))
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/inventory.rs:39]: As above, worth looking at in the future, but will require a few changes in a follow-on PR.

File: crates/volta-core/src/inventory.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (chriskrycho) [reply crates/volta-core/src/inventory.rs:29]: I wondered about that; totally makes sense. Thanks!

**Pull request merged**